File: adventjs/2025/17/solution.js
/**
 * @param {string[][]} board
 * @returns {boolean}
 */
function hasFourLights(board) {
  if (board.length === 0) return false;
  
  const rows = board.length;
  const cols = board[0].length;
  
  // Check horizontal lines
  for (let r = 0; r < rows; r++) {
    let count = 0;
    let currentColor = null;
    
    for (let c = 0; c < cols; c++) {
      const cell = board[r][c];
      if (cell === '.') {
        count = 0;
        currentColor = null;
      } else if (cell === currentColor) {
        count++;
        if (count >= 4) return true;
      } else {
        count = 1;
        currentColor = cell;
        if (count >= 4) return true;
      }
    }
  }
  
  // Check vertical lines
  for (let c = 0; c < cols; c++) {
    let count = 0;
    let currentColor = null;
    
    for (let r = 0; r < rows; r++) {
      const cell = board[r][c];
      if (cell === '.') {
        count = 0;
        currentColor = null;
      } else if (cell === currentColor) {
        count++;
        if (count >= 4) return true;
      } else {
        count = 1;
        currentColor = cell;
        if (count >= 4) return true;
      }
    }
  }
  
  return false;
}

File: adventjs/2025/17/solution.py
def has_four_lights(board: list[list[str]]) -> bool:
  if not board:
    return False
  
  def check_line(cells):
    count, color = 0, None
    for cell in cells:
      if cell == '.':
        count, color = 0, None
      elif cell == color:
        count += 1
      else:
        count, color = 1, cell
      if count >= 4:
        return True
    return False
  
  rows, cols = len(board), len(board[0])
  return any(check_line(board[r]) for r in range(rows)) or any(check_line([board[r][c] for r in range(rows)]) for c in range(cols))
